### ISSUE TITLE
helm: teardown of shard causes the vttablet pods to hang

### ISSUE THREAD

**Issue by sougou** [vitessio/vitess#6044]
@derekperkins @carsonoid @enisoc 

I was running through the steps of the helm example. The following step causes the master pod to hang. I believe it's because we try to reparent, but there's no vttablet to reparent to.

Here are the log messages of the vttablet when the kill pod command is issued:
```
I0410 05:53:21.175919       1 rpc_replication.go:375] DemoteMaster disabling query service
I0410 05:53:21.176053       1 query.go:73] exec SET GLOBAL read_only = ON
I0410 05:53:21.176416       1 rpc_server.go:95] TabletManager.DemoteMaster()(on zone1-0528462200 from ): (*tabletmanagerdata.DemoteMasterResponse)(nil)
I0410 05:53:51.184912       1 query.go:73] exec SET GLOBAL read_only = OFF
I0410 05:53:51.188455       1 rpc_replication.go:453] UndoDemoteMaster re-enabling query service
I0410 05:53:51.188514       1 tabletserver.go:300] TabletServer state: NOT_SERVING (Not Serving) -> NOT_SERVING (Transitioning)
I0410 05:53:51.188532       1 tx_pool.go:139] Starting transaction id: {1586497131716270412}
I0410 05:53:51.188609       1 tabletserver.go:300] TabletServer state: NOT_SERVING (Transitioning) -> SERVING
I0410 05:53:51.188652       1 rpc_server.go:95] TabletManager.UndoDemoteMaster()(on zone1-0528462200 from ): (*tabletmanagerdata.UndoDemoteMasterResponse)(nil)
I0410 05:53:53.831500       1 healthcheck.go:270] Disabling query service because of health-check failure: <nil>
I0410 05:53:53.831832       1 tabletserver.go:300] TabletServer state: SERVING -> SHUTTING_DOWN (Shutting Down)
I0410 05:53:53.832344       1 tx_engine.go:435] No grace period specified: performing normal wait.
I0410 05:53:53.832540       1 tabletserver.go:300] TabletServer state: SHUTTING_DOWN (Shutting Down) -> NOT_SERVING (Not Serving)
```

And later, when I delete the shard from topo, I see the following messages, but the pod remains in `Terminating` state forever:

```
I0410 05:56:00.477972       1 shard_sync.go:84] Change in shard record
E0410 05:56:00.477999       1 shard_sync.go:87] Shard watch failed: node doesn't exist: vt-12956164405595433652
I0410 05:56:00.478005       1 shard_watcher.go:56] Stopping shard watch...
I0410 05:56:00.478009       1 shard_watcher.go:62] Shard watch stopped.
E0410 05:56:00.480811       1 shard_sync.go:107] Failed to sync shard record: node doesn't exist: keyspaces/customer/shards/0/Shard
```

**Comment by enisoc** [comment]
As far as I can tell, we give the master Pod an effectively unlimited amount of time to reparent. I don't know whether there's a prescribed way to tear down a cluster deployed by helm to avoid getting into this state, but once you're in it I would try deleting the stuck Pod with `--grace-period=1`.

**Comment by derekperkins** [comment]
Yep. The default is to not let a master willingly go down, which has definitely saved us before. It's the right production setting, but I don't know that there's anything to make cleanup easier without manually deleting each pod with `--grace-period=1`

**Comment by sougou** [comment]
The problem with helm right now is that this is done implicitly by the fact that we push a new helm config with the shards removed. We could change our instructions to ask them to do a force delete, and explain why.

How is this handled in the operator?

**Comment by derekperkins** [comment]
Not sure about the operator, but at the end of the day, helm is just too blunt of an instrument to actually manage Vitess. That's why we ended up just using helm to generate the base yaml and have managed those manually since then. We'll migrate to the operator later this year when we have a chance.

**Comment by enisoc** [comment]
> How is this handled in the operator?

Our operator doesn't use preStop hooks like the Helm chart does. If you want to take down a master gracefully, you have to first annotate the Pod with `drain.planetscale.com/started` and then wait for the operator to add `drain.planetscale.com/finished` to indicate it has reparented away before you delete the Pod.

This explicit protocol separates the concerns of graceful draining and deletion. You are free to delete without draining if that's what you want, such as during tear-down. The trade-off is that our operator's answer to, "What if someone ignores the drain protocol and deletes a master Pod without notice?" is currently to hope you're running on remote disk and wait for the master to reschedule somewhere else and perform crash recovery.

**Comment by sougou** [comment]
I was able to get the example to behave better with this change:
```
diff --git a/helm/vitess/templates/_vttablet.tpl b/helm/vitess/templates/_vttablet.tpl
index af9d2e62c..08ad25b29 100644
--- a/helm/vitess/templates/_vttablet.tpl
+++ b/helm/vitess/templates/_vttablet.tpl
@@ -130,7 +130,7 @@ spec:
         type: {{ $tablet.type | quote }}
     spec:
       serviceAccountName: vttablet
-      terminationGracePeriodSeconds: 60000000
+      terminationGracePeriodSeconds: 1
 {{ include "pod-security" . | indent 6 }}
 {{ include "vttablet-affinity" (tuple $cellClean $keyspaceClean $shardClean $cell.region) | indent 6 }}
```

I'm now debating between parameterizing this setting for the sake of the example vs. adding an extra step to force-delete pods. I feel like parameterization is better, because I can see situations where users may want to fine-tune this.

**Comment by derekperkins** [comment]
Nothing against adding a parameter. I was unprepared for the explosion of parameters in the helm charts, and so adding new parameters is a much bigger pain than it should be.

**Comment by carsonoid** [comment]
You need to be careful when changing that setting.
`terminationGracePeriodSeconds`
tells the Kubelet how long to wait for a container to stop gracefully
before it forcefully kills it. Basically, it's a countdown to a `kill -9`.
So 1 second is probably way too short since it will make it very likely
that other containers in the pod, like mysql, will be force-killed and may
not give them time to do their shutdown tasks. But adding an override
option is fine since it puts that decision and its consequences on the
end-user.

**Comment by sougou** [comment]
Fixed via #6048
I've added explicit comments in the overrides that it's not recommended for prod use.
